agent/tcs/handler/handler.go

```diff
@@ -91,15 +90,18 @@ func startTelemetrySession(params TelemetrySessionParams, statsEngine stats.Engi
 	}
 	log.Debug("Connecting to TCS endpoint " + tcsEndpoint)
 	url := formatURL(tcsEndpoint, params.Cfg.Cluster, params.ContainerInstanceArn)
-	return startSession(url, params.Cfg.AWSRegion, params.CredentialProvider, params.AcceptInvalidCert, statsEngine, defaultPublishMetricsInterval, params.DeregisterInstanceStream)
+	return startSession(url, params.Cfg.AWSRegion, params.CredentialProvider, params.AcceptInvalidCert, statsEngine, defaultPublishMetricsInterval, params.DeregisterInstanceEventStream)
 }
 
-func startSession(url string, region string, credentialProvider *credentials.Credentials, acceptInvalidCert bool, statsEngine stats.Engine, publishMetricsInterval time.Duration, deregisterInstanceStream *event.ACSDeregisterInstanceStream) error {
+func startSession(url string, region string, credentialProvider *credentials.Credentials, acceptInvalidCert bool, statsEngine stats.Engine, publishMetricsInterval time.Duration, deregisterInstanceEventStream *eventstream.EventStream) error {
 	client := tcsclient.New(url, region, credentialProvider, acceptInvalidCert, statsEngine, publishMetricsInterval)
 	defer client.Close()
 
-	deregisterInstanceStream.Subscribe(client.Disconnect)
-	defer deregisterInstanceStream.Unsubscribe(client.Disconnect)
+	err := deregisterInstanceEventStream.Subscribe(deregisterContainerInstanceHandler, client.Disconnect)
+	if err != nil {
+		return err
```

Comment:
This returned err will be logged above, so there won't be duplicate info in the log file.